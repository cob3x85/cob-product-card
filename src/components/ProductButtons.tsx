import React, { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css'

export interface Props {
    className?: string;
    style?: React.CSSProperties
}

export const ProductButtons = ({ className, style }: Props) => {

    const { increaseBy, counter, maxCount } = useContext(ProductContext);

    // const isMaxReached = useCallback(
    //     () => {
    //         if (maxCount) {
    //             return counter >= maxCount
    //         }
    //         return false
    //     },
    //     [counter, maxCount],
    // );

    // One liner example
    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount,
        [counter, maxCount],
    );


    // Nuevo callback  usMaxreached = useCallback, cuando el [counter, maxCount] cambie
    // TRUE si el counter == maxCount
    // FALSE si no lo es
    // si es max reached entonces cambiar el estilo del boton mas

    return (
        <div
            className={`${styles.buttonsContainer} ${className}`}
            style={style}
        >
            <button
                className={styles.buttonMinus}
                onClick={() => increaseBy(-1)}> - </button>

            <div className={styles.countLabel}> {counter} </div>

            <button
                disabled={isMaxReached()}
                className={isMaxReached() ? styles.disabled : styles.buttonAdd}
                onClick={() => increaseBy(+1)}> + </button>
        </div>
    );
}
