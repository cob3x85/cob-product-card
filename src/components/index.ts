import { ProductCardHOCProps } from '../interfaces/interfaces';
import { ProductCard as ProductCardHOC } from './ProductCard';

import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductTitle } from './ProductTitle';

export { ProductTitle } from './ProductTitle';
export { ProductImage } from "./ProductImage";
export { ProductButtons } from './ProductButtons';

// Object assign permite extender las propiedades del objecto
// nos sirve para usar el Compound Component Pattern
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})

export default ProductCard;

