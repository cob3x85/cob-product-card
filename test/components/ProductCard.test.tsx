import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product1 } from '../data/product';

const { act } = renderer

describe('ProductCard', () => {
    test('debe mostrar el componente correctamente', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    () => (
                        <h1>Card</h1>
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('debe de incrementar el counter', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}>
                {
                    ({ count, increaseBy }) => (
                        <>
                            <h1>Card</h1>
                            <span>{count}</span>
                            <button onClick={() => increaseBy(1)}>Add</button>
                        </>
                    )
                }
            </ProductCard>
        );
        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
        //  When testing, code that causes React state updates should be wrapped into act(...):
        act(() => {
            (tree as any).children[2].props.onClick();
        });
        tree = wrapper.toJSON()
        expect((tree as any).children[1].children[0]).toBe('1')
    });
})
