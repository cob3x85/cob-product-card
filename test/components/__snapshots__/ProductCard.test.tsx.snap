// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard debe de incrementar el counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;

exports[`ProductCard debe mostrar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Card
  </h1>
</div>
`;
